src/main/kotlin/com/openlattice/data/ids/jobs/FixDuplicateIdAssignmentJob.kt

```diff
@@ -0,0 +1,301 @@
+import com.geekbeast.rhizome.jobs.AbstractDistributedJob
+import com.geekbeast.rhizome.jobs.JobStatus
+import com.openlattice.client.serialization.SerializationConstants
+import com.openlattice.data.EntityKey
+import com.openlattice.graph.partioning.REPARTITION_DATA_COLUMNS
+import com.openlattice.graph.partioning.REPARTITION_EDGES_COLUMNS
+import com.openlattice.graph.partioning.REPARTITION_IDS_COLUMNS
+import com.openlattice.hazelcast.serializers.decorators.IdGenerationAware
+import com.openlattice.hazelcast.serializers.decorators.MetastoreAware
+import com.openlattice.ids.HazelcastIdGenerationService
+import com.openlattice.postgres.PostgresArrays
+import com.openlattice.postgres.PostgresColumn.*
+import com.openlattice.postgres.PostgresTable.*
+import com.openlattice.postgres.ResultSetAdapters
+import com.openlattice.postgres.streams.BasePostgresIterable
+import com.openlattice.postgres.streams.PreparedStatementHolderSupplier
+import com.zaxxer.hikari.HikariDataSource
+import java.lang.IllegalStateException
+import java.sql.PreparedStatement
+import java.util.*
+
+const val BATCH_SIZE = 128000
+
+/**
+ *
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+class FixDuplicateIdAssignmentJob(
+        state: FixDuplicateIdAssignmentJobState
+) : AbstractDistributedJob<Long, FixDuplicateIdAssignmentJobState>(state), MetastoreAware, IdGenerationAware {
+    companion object {
+        @Transient
+        private val mapper = ObjectMappers.newJsonMapper()
+    }
+
+    @JsonCreator
+    constructor(
+            id: UUID?,
+            taskId: Long?,
+            status: JobStatus,
+            progress: Byte,
+            hasWorkRemaining: Boolean,
+            result: Long?,
+            state: FixDuplicateIdAssignmentJobState
+    ) : this(state) {
+        initialize(id, taskId, status, progress, hasWorkRemaining, result)
+    }
+
+    @Transient
+    private lateinit var hds: HikariDataSource
+
+    @Transient
+    private lateinit var idService: HazelcastIdGenerationService
+
+    override fun processNextBatch() {
+        /**
+         * Process ids in batches.
+         */
+        val ids = BasePostgresIterable(PreparedStatementHolderSupplier(hds, SELECT_NEXT_IDS) { ps ->
+            ps.setObject(1, state.id)
+        }) { rs ->
+            ResultSetAdapters.id(rs)
+        }.toList()
+
+        val entityKeysAndIds = BasePostgresIterable(PreparedStatementHolderSupplier(hds, SELECT_NEXT_DUPLICATES) { ps ->
+            ps.setArray(1, PostgresArrays.createUuidArray(ps.connection, ids))
+        }) { rs ->
+            ResultSetAdapters.id(rs) to mapper.readValue<List<EntityKey>>(rs.getString(DUPLICATES_FIELD))
+        }.toMap()
+
+        entityKeysAndIds
+                .asSequence()
+                .forEach { (id, entityKeys) ->
+                    val entitySetAppearances = entityKeys.groupingBy { it.entitySetId }.eachCount()
+
+                    if (entitySetAppearances.values.all { it == 1 }) {
```

Comment:
This seems doable on a per-entity set basis, yeah? so if you have
```
{
  entitySet1: 1,
  entitySet2: 1,
  entitySet3: 2
}
```
then you could assign new ids for the entities in `entitySet1` and `entitySet2` and only have to manually address the collisions from `entitySet3`
(if I'm reading this correctly)

Reply:
Yup. Good catch.